## deepTS
- deepTS is a web-based Galaxy platform exploring transcriptional switch events from large-scale RNA sequencing data.
- deepTS consists of three main functional modules, covering the processes of read cleaning and mapping; transcriptome map construction; expression abundance estimation; and TS identification, visualization, and exploration. In addition, deepTS provides a series of mutually independent, end-to-end functions, including multiple-algorithm expression abundance estimation (e.g., StringTie, salmon), multiple-condition TS analysis pipelines, multiple-level TS characterization, and multiple-form visualization, to allow users to perform TS analysis using either raw RNA-Seq data or expression abundance matrices directly.
- deepTS project is hosted on GitHub (https://github.com/cma2015/deepTS), and can be accessed from http://deepTS.nwafu.edu.cn. The deepTS Docker images is available at XXX.

    ![deepTS](https://github.com/chenzhuod/deepTS/Tutorials/deepTS_images/Figure 1_Overview_20191218.png)

## Overview of functional modules in deepTS

## How to use deepTS

- Tutorials for deepTS: XXX
- Test datasets referred in the tutorials for deepTS: XXX

## News and updates

### deepTS released on Dec XXX, 2019

- deepTS source codes, web server and Docker image were released for the first time.

## How to access help
* For any bugs/issues, please feel free to leave a message at Github [issues](<https://github.com/cma2015/deepTS/issues>). We will try our best to deal with all issues as soon as possible.
* For any suggestions/comments, please send emails to: __Siyuan Chen__ <chenzhuod@gmail.com> or __Zhixu Qiu__ <zhixuqiu2015@gmail.com>

## How to cite this work
Zhixu Qiu, Siyuan Chen, Yuhong Qi,  Chunni Liu, Xiaorong Zhang, Shang Xie, Jingjing Zhai, Chuang Ma, Exploring transcriptional switch events from large-scale RNA sequencing data using deepTS (submitted)
