# [Identification of TS events ](https://github.com/cma2015/DeepTS/blob/master/Tutorials/Test_data/Module2)

## Pairwise transcriptome comparison
- **Genome annotation file** is in the directory named `pairwiseTS_annotation_file.gtf`
- **Transcript expression abundance matrix** is in the directorynamed `pairwiseTS_transcript_expression.tsv`
- **Experiment design matrix** is in the directory named `pairwiseTS_experiment_design.tsv`

- **Gene-transcript ID matrix** is in the directory named `gene_transcriptID.tsv` 
The first, second and third columns are transcripts, genes and corresponding chromosomes. Chromosomes are in digital format.

## Time-series transcriptome comparison

- **Transcript expression abundance matrix** is in the directory named `timeSeriesTS_transcript_expression.tsv`
- **Experiment design matrix*** is in the directory named `timeSeriesTS_experiment_design.tsv`
- **Gene-transcript ID matrix** is in the directory named `gene_transcriptID.tsv`

## Population transcriptome comparison
- **Genotype data** is in the directory named `popultaionTS_genotype_hmpfile.tsv`
the chromsome names in hmp file are in digital format. 
- **Transcript expression abundance matrix** is in the directory named `popultaionTS_transcript_expression.tsv`
- **Gene-transcript ID matrix** is in the directory named `gene_transcriptID.tsv`

